packages/tkg-clusterclass-aws/bundle/config/upstream/base.yaml

```
@@ -273,6 +273,11 @@ spec:
       openAPIV3Schema:
         type: object
         properties: {}
+  - name: controlPlaneCertificateRotationBefore
+    required: false
+    schema:
+      openAPIV3Schema:
+        type: integer
```

```suggestion
        type: integer
        minimum: 7
```
Might make sense to set a minimum here. Inclusive minimum in Cluster API is 7 
(https://github.com/kubernetes-sigs/cluster-api/blob/650b1879a586228d9ec5d81f3dd86016f588aa5d/controlplane/kubeadm/api/v1beta1/kubeadm_control_plane_webhook.go#L339)

---

Same in other places

---

We also wanted a default of 90 as per the TEP for classy clusters, and opt-in for existing YTT based clusters.

---

Yes. But if users want to disable certificate automatic renewal, how should users do? Now, if users don't set the variable, the cluster is created without certificate automatic renewal.

---

That's a good point. If we have a default folks cannot just simply "unset" it. We could use 0 or -1 as a way to opt-out

---

Now I have updated the logic and It used default value 90.
1. Users don't set the value: they will create the cluster enabling certificate automatic renewal with default value 90
2. Users set the value with the number less than 1: they will create the cluster disabling certificate automatic renewal
3. User set the value with the integer greater than 1: they will create the cluster enabling certificate automatic renewal with the input value

---

Probably the best tradeoff if we want to encode this behavior in just one single basic variable.

Only issue is that every value < 7 and >= 1 will be rejected by the KubeadmControlPlane webhook as the supported minimum there is 7

---

> Probably the best tradeoff if we want to encode this behavior in just one single basic variable.
> 
> Only issue is that every value < 7 and >= 1 will be rejected by the KubeadmControlPlane webhook as the supported minimum there is 7
> 
> We could validate here but that would require us setting `minimum: 7` which makes it impossible to use < 1 as a signal for disabled.

Based on it, we should throw validation error if the value is set in [1, 7).

---

Yup, Unfortunately that's something you cannot express with the current Open API schema validation (there's only min/max allowed values).

---

Yes. I cannot use minimum annotation for we support -1/0. I will update it using function checking.